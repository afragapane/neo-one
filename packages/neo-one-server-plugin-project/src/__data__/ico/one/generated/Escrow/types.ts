/* @hash 912635047a992ae37d892936e4acfb6d */
// tslint:disable
/* eslint-disable */
import {
  AddressString,
  Client,
  ForwardValue,
  GetOptions,
  InvocationTransaction,
  InvokeReceipt,
  SmartContract,
  TransactionOptions,
  TransactionResult,
} from '@neo-one/client';
import BigNumber from 'bignumber.js';

export type EscrowEvent = never;

export interface EscrowSmartContract<TClient extends Client = Client> extends SmartContract<TClient, EscrowEvent> {
  readonly approveReceiveTransfer: {
    (
      from: AddressString,
      amount: BigNumber,
      asset: AddressString,
      to: AddressString,
      options?: TransactionOptions,
    ): Promise<TransactionResult<InvokeReceipt<boolean, EscrowEvent>, InvocationTransaction>>;
    readonly confirmed: {
      (
        from: AddressString,
        amount: BigNumber,
        asset: AddressString,
        to: AddressString,
        options?: TransactionOptions & GetOptions,
      ): Promise<InvokeReceipt<boolean, EscrowEvent> & { readonly transaction: InvocationTransaction }>;
    };
  };
  readonly forwardApproveReceiveTransferArgs: (to: AddressString) => [ForwardValue];
  readonly balanceOf: (from: AddressString, to: AddressString, asset: AddressString) => Promise<BigNumber>;
  readonly claim: {
    (
      from: AddressString,
      to: AddressString,
      asset: AddressString,
      amount: BigNumber,
      options?: TransactionOptions,
    ): Promise<TransactionResult<InvokeReceipt<boolean, EscrowEvent>, InvocationTransaction>>;
    readonly confirmed: {
      (
        from: AddressString,
        to: AddressString,
        asset: AddressString,
        amount: BigNumber,
        options?: TransactionOptions & GetOptions,
      ): Promise<InvokeReceipt<boolean, EscrowEvent> & { readonly transaction: InvocationTransaction }>;
    };
  };
  readonly deploy: {
    (options?: TransactionOptions): Promise<
      TransactionResult<InvokeReceipt<boolean, EscrowEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (options?: TransactionOptions & GetOptions): Promise<
        InvokeReceipt<boolean, EscrowEvent> & { readonly transaction: InvocationTransaction }
      >;
    };
  };
  readonly onRevokeSendTransfer: {
    (from: AddressString, amount: BigNumber, asset: AddressString, options?: TransactionOptions): Promise<
      TransactionResult<InvokeReceipt<undefined, EscrowEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (
        from: AddressString,
        amount: BigNumber,
        asset: AddressString,
        options?: TransactionOptions & GetOptions,
      ): Promise<InvokeReceipt<undefined, EscrowEvent> & { readonly transaction: InvocationTransaction }>;
    };
  };
  readonly refund: {
    (
      from: AddressString,
      to: AddressString,
      asset: AddressString,
      amount: BigNumber,
      options?: TransactionOptions,
    ): Promise<TransactionResult<InvokeReceipt<boolean, EscrowEvent>, InvocationTransaction>>;
    readonly confirmed: {
      (
        from: AddressString,
        to: AddressString,
        asset: AddressString,
        amount: BigNumber,
        options?: TransactionOptions & GetOptions,
      ): Promise<InvokeReceipt<boolean, EscrowEvent> & { readonly transaction: InvocationTransaction }>;
    };
  };
}
