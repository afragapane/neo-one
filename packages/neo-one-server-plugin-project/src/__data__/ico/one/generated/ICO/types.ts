/* @hash 6811f11b9e627cc6bba5601bf203e9dd */
// tslint:disable
/* eslint-disable */
import {
  AddressString,
  Event,
  GetOptions,
  Hash256String,
  InvocationTransaction,
  InvokeReceipt,
  InvokeReceiveTransactionOptions,
  InvokeSendTransactionOptions,
  ReadSmartContract,
  SmartContract,
  TransactionOptions,
  TransactionResult,
} from '@neo-one/client';
import BigNumber from 'bignumber.js';

export type ICOEvent = ICOTransferEvent | ICOApproveSendTransferEvent | ICORevokeSendTransferEvent;

export interface ICOTransferEventParameters {
  readonly from: AddressString | undefined;
  readonly to: AddressString | undefined;
  readonly amount: BigNumber;
}
export interface ICOTransferEvent extends Event<'transfer', ICOTransferEventParameters> {}
export interface ICOApproveSendTransferEventParameters {
  readonly from: AddressString;
  readonly to: AddressString;
  readonly amount: BigNumber;
}
export interface ICOApproveSendTransferEvent
  extends Event<'approveSendTransfer', ICOApproveSendTransferEventParameters> {}
export interface ICORevokeSendTransferEventParameters {
  readonly from: AddressString;
  readonly to: AddressString;
  readonly amount: BigNumber;
}
export interface ICORevokeSendTransferEvent extends Event<'revokeSendTransfer', ICORevokeSendTransferEventParameters> {}

export interface ICOSmartContract extends SmartContract<ICOReadSmartContract> {
  readonly amountPerNEO: () => Promise<BigNumber>;
  readonly deploy: {
    (
      owner?: AddressString,
      startTimeSeconds?: BigNumber,
      icoDurationSeconds?: BigNumber,
      options?: TransactionOptions,
    ): Promise<TransactionResult<InvokeReceipt<boolean, ICOEvent>, InvocationTransaction>>;
    readonly confirmed: {
      (
        owner?: AddressString,
        startTimeSeconds?: BigNumber,
        icoDurationSeconds?: BigNumber,
        options?: TransactionOptions & GetOptions,
      ): Promise<InvokeReceipt<boolean, ICOEvent> & { readonly transaction: InvocationTransaction }>;
    };
  };
  readonly icoDurationSeconds: () => Promise<BigNumber>;
  readonly mintTokens: {
    (options?: InvokeReceiveTransactionOptions): Promise<
      TransactionResult<InvokeReceipt<boolean, ICOEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (options?: InvokeReceiveTransactionOptions & GetOptions): Promise<
        InvokeReceipt<boolean, ICOEvent> & { readonly transaction: InvocationTransaction }
      >;
    };
  };
  readonly owner: () => Promise<AddressString>;
  readonly refundAssets: {
    (transactionHash: Hash256String, options?: InvokeSendTransactionOptions): Promise<
      TransactionResult<InvokeReceipt<boolean, ICOEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (transactionHash: Hash256String, options?: InvokeSendTransactionOptions & GetOptions): Promise<
        InvokeReceipt<boolean, ICOEvent> & { readonly transaction: InvocationTransaction }
      >;
    };
  };
  readonly remaining: () => Promise<BigNumber>;
  readonly startTimeSeconds: () => Promise<BigNumber>;
}

export interface ICOReadSmartContract extends ReadSmartContract<ICOEvent> {
  readonly amountPerNEO: () => Promise<BigNumber>;
  readonly icoDurationSeconds: () => Promise<BigNumber>;
  readonly owner: () => Promise<AddressString>;
  readonly remaining: () => Promise<BigNumber>;
  readonly startTimeSeconds: () => Promise<BigNumber>;
}
