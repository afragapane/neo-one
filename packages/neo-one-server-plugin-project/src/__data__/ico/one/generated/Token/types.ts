/* @hash f18e529fec9346fe9495de5c7dcf153a */
// tslint:disable
/* eslint-disable */
import {
  AddressString,
  Event,
  ForwardValue,
  GetOptions,
  Hash256String,
  InvocationTransaction,
  InvokeReceipt,
  InvokeSendTransactionOptions,
  ReadSmartContract,
  SmartContract,
  TransactionOptions,
  TransactionResult,
} from '@neo-one/client';
import BigNumber from 'bignumber.js';

export type TokenEvent = TokenTransferEvent | TokenApproveSendTransferEvent | TokenRevokeSendTransferEvent;

export interface TokenTransferEventParameters {
  readonly from: AddressString | undefined;
  readonly to: AddressString | undefined;
  readonly amount: BigNumber;
}
export interface TokenTransferEvent extends Event<'transfer', TokenTransferEventParameters> {}
export interface TokenApproveSendTransferEventParameters {
  readonly from: AddressString;
  readonly to: AddressString;
  readonly amount: BigNumber;
}
export interface TokenApproveSendTransferEvent
  extends Event<'approveSendTransfer', TokenApproveSendTransferEventParameters> {}
export interface TokenRevokeSendTransferEventParameters {
  readonly from: AddressString;
  readonly to: AddressString;
  readonly amount: BigNumber;
}
export interface TokenRevokeSendTransferEvent
  extends Event<'revokeSendTransfer', TokenRevokeSendTransferEventParameters> {}

export interface TokenSmartContract extends SmartContract<TokenReadSmartContract> {
  readonly approveReceiveTransfer: {
    (from: AddressString, amount: BigNumber, asset: AddressString, options?: TransactionOptions): Promise<
      TransactionResult<InvokeReceipt<boolean, TokenEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (
        from: AddressString,
        amount: BigNumber,
        asset: AddressString,
        options?: TransactionOptions & GetOptions,
      ): Promise<InvokeReceipt<boolean, TokenEvent> & { readonly transaction: InvocationTransaction }>;
    };
  };
  readonly approveSendTransfer: {
    (from: AddressString, to: AddressString, amount: BigNumber, options?: TransactionOptions): Promise<
      TransactionResult<InvokeReceipt<boolean, TokenEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (from: AddressString, to: AddressString, amount: BigNumber, options?: TransactionOptions & GetOptions): Promise<
        InvokeReceipt<boolean, TokenEvent> & { readonly transaction: InvocationTransaction }
      >;
    };
  };
  readonly approvedTransfer: (from: AddressString, to: AddressString) => Promise<BigNumber>;
  readonly balanceOf: (address: AddressString) => Promise<BigNumber>;
  readonly decimals: () => Promise<BigNumber>;
  readonly deploy: {
    (options?: TransactionOptions): Promise<
      TransactionResult<InvokeReceipt<boolean, TokenEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (options?: TransactionOptions & GetOptions): Promise<
        InvokeReceipt<boolean, TokenEvent> & { readonly transaction: InvocationTransaction }
      >;
    };
  };
  readonly issue: {
    (to: AddressString, amount: BigNumber, options?: TransactionOptions): Promise<
      TransactionResult<InvokeReceipt<boolean, TokenEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (to: AddressString, amount: BigNumber, options?: TransactionOptions & GetOptions): Promise<
        InvokeReceipt<boolean, TokenEvent> & { readonly transaction: InvocationTransaction }
      >;
    };
  };
  readonly name: () => Promise<string>;
  readonly onRevokeSendTransfer: {
    (from: AddressString, amount: BigNumber, asset: AddressString, options?: TransactionOptions): Promise<
      TransactionResult<InvokeReceipt<undefined, TokenEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (
        from: AddressString,
        amount: BigNumber,
        asset: AddressString,
        options?: TransactionOptions & GetOptions,
      ): Promise<InvokeReceipt<undefined, TokenEvent> & { readonly transaction: InvocationTransaction }>;
    };
  };
  readonly refundAssets: {
    (transactionHash: Hash256String, options?: InvokeSendTransactionOptions): Promise<
      TransactionResult<InvokeReceipt<boolean, TokenEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (transactionHash: Hash256String, options?: InvokeSendTransactionOptions & GetOptions): Promise<
        InvokeReceipt<boolean, TokenEvent> & { readonly transaction: InvocationTransaction }
      >;
    };
  };
  readonly revokeSendTransfer: {
    (from: AddressString, to: AddressString, amount: BigNumber, options?: TransactionOptions): Promise<
      TransactionResult<InvokeReceipt<boolean, TokenEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (from: AddressString, to: AddressString, amount: BigNumber, options?: TransactionOptions & GetOptions): Promise<
        InvokeReceipt<boolean, TokenEvent> & { readonly transaction: InvocationTransaction }
      >;
    };
  };
  readonly symbol: () => Promise<string>;
  readonly totalSupply: () => Promise<BigNumber>;
  readonly transfer: {
    (
      from: AddressString,
      to: AddressString,
      amount: BigNumber,
      options?: TransactionOptions,
      ...approveArgs: ForwardValue[]
    ): Promise<TransactionResult<InvokeReceipt<boolean, TokenEvent>, InvocationTransaction>>;
    (from: AddressString, to: AddressString, amount: BigNumber, ...approveArgs: ForwardValue[]): Promise<
      TransactionResult<InvokeReceipt<boolean, TokenEvent>, InvocationTransaction>
    >;
    readonly confirmed: {
      (
        from: AddressString,
        to: AddressString,
        amount: BigNumber,
        options?: TransactionOptions & GetOptions,
        ...approveArgs: ForwardValue[]
      ): Promise<InvokeReceipt<boolean, TokenEvent> & { readonly transaction: InvocationTransaction }>;
      (from: AddressString, to: AddressString, amount: BigNumber, ...approveArgs: ForwardValue[]): Promise<
        InvokeReceipt<boolean, TokenEvent> & { readonly transaction: InvocationTransaction }
      >;
    };
  };
}

export interface TokenReadSmartContract extends ReadSmartContract<TokenEvent> {
  readonly approvedTransfer: (from: AddressString, to: AddressString) => Promise<BigNumber>;
  readonly balanceOf: (address: AddressString) => Promise<BigNumber>;
  readonly decimals: () => Promise<BigNumber>;
  readonly name: () => Promise<string>;
  readonly symbol: () => Promise<string>;
  readonly totalSupply: () => Promise<BigNumber>;
}
