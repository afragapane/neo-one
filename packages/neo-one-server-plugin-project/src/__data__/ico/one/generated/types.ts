/* @hash 4fe61ed1ec155170b977df3927670c05 */
// tslint:disable
/* eslint-disable */
import { ICOSmartContract } from './ICO/types';
import { TokenSmartContract } from './Token/types';

export interface Contracts {
  readonly ico: ICOSmartContract;
  readonly token: TokenSmartContract;
}
