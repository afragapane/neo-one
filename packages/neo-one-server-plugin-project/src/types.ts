import * as path from 'path';

export interface BuildTaskListOptions {
  readonly command: 'build';
  readonly rootDir: string;
}

export interface ResetTaskListOptions {
  readonly command: 'reset';
  readonly projectID?: string;
  readonly rootDir?: string;
}

export type ExecuteTaskListOptions = BuildTaskListOptions | ResetTaskListOptions;

export interface NetworkRequestOptions {
  readonly type: 'network';
  readonly projectID: string;
}

export interface SourceMapsRequestOptions {
  readonly type: 'sourceMaps';
  readonly projectID: string;
}

export type RequestOptions = NetworkRequestOptions | SourceMapsRequestOptions;

export interface ProjectConfig {
  readonly paths: {
    readonly contracts: string;
    readonly generated: string;
  };
  readonly codegen: {
    readonly javascript: boolean;
  };
}

export const projectConfigSchema = {
  paths: {
    contracts: {
      format: String,
      default: path.join('one', 'contracts'),
    },
    generated: {
      format: String,
      default: path.join('one', 'generated'),
    },
  },
  codegen: {
    javascript: {
      format: Boolean,
      default: false,
    },
  },
};
