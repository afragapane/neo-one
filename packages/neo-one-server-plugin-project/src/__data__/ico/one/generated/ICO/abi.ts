// tslint:disable
import { ABI } from '@neo-one/client';

export const iCOABI: ABI = {
  events: [
    {
      name: 'transfer',
      parameters: [
        {
          name: 'from',
          optional: true,
          type: 'Hash160',
        },
        {
          name: 'to',
          optional: true,
          type: 'Hash160',
        },
        {
          decimals: 8,
          name: 'amount',
          optional: false,
          type: 'Integer',
        },
      ],
    },
    {
      name: 'refund',
      parameters: [],
    },
    {
      name: 'trace',
      parameters: [
        {
          decimals: 0,
          name: 'line',
          type: 'Integer',
        },
      ],
    },
    {
      name: 'error',
      parameters: [
        {
          decimals: 0,
          name: 'line',
          type: 'Integer',
        },
        {
          name: 'message',
          type: 'String',
        },
      ],
    },
    {
      name: 'console.log',
      parameters: [
        {
          decimals: 0,
          name: 'line',
          type: 'Integer',
        },
        {
          name: 'args',
          type: 'ByteArray',
        },
      ],
    },
  ],
  functions: [
    {
      constant: true,
      name: 'name',
      returnType: {
        optional: false,
        type: 'String',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'symbol',
      returnType: {
        optional: false,
        type: 'String',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'decimals',
      returnType: {
        decimals: 0,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'amountPerNEO',
      returnType: {
        decimals: 0,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'owner',
      returnType: {
        optional: false,
        type: 'Hash160',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'startTimeSeconds',
      returnType: {
        decimals: 0,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'icoDurationSeconds',
      returnType: {
        decimals: 0,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'totalSupply',
      returnType: {
        decimals: 8,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'balanceOf',
      parameters: [
        {
          name: 'address',
          optional: false,
          type: 'Hash160',
        },
      ],
      returnType: {
        decimals: 8,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: false,
      name: 'transfer',
      parameters: [
        {
          name: 'from',
          optional: false,
          type: 'Hash160',
        },
        {
          name: 'to',
          optional: false,
          type: 'Hash160',
        },
        {
          decimals: 8,
          name: 'amount',
          optional: false,
          type: 'Integer',
        },
      ],
      returnType: {
        optional: false,
        type: 'Boolean',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'remaining',
      returnType: {
        decimals: 0,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: false,
      name: 'mintTokens',
      parameters: [],
      returnType: {
        optional: false,
        type: 'Boolean',
      },
      verify: true,
    },
    {
      constant: false,
      name: 'deploy',
      parameters: [
        {
          name: 'owner',
          optional: true,
          type: 'Hash160',
        },
        {
          decimals: 0,
          name: 'startTimeSeconds',
          optional: true,
          type: 'Integer',
        },
        {
          decimals: 0,
          name: 'icoDurationSeconds',
          optional: true,
          type: 'Integer',
        },
      ],
      returnType: {
        type: 'Boolean',
      },
      verify: false,
    },
  ],
};
