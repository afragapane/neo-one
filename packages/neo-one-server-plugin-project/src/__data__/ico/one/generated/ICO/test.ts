// tslint:disable
import { setupContractTest, SetupTestResult } from '@neo-one/smart-contract-compiler';
import * as path from 'path';
import { ICOSmartContract } from './types';

export const setupICOTest = async (): Promise<SetupTestResult<ICOSmartContract>> =>
  setupContractTest<ICOSmartContract>(path.resolve(__dirname, '../../contracts/ICO.ts'), 'ICO');
