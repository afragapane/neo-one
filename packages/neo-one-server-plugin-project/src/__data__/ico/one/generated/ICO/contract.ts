/* @hash 89f96ab5ffc5809c1c00abe42c50c8be */
// tslint:disable
/* eslint-disable */
import { Client, SmartContractDefinition } from '@neo-one/client';
import { icoABI } from './abi';
import { ICOSmartContract } from './types';
import { sourceMaps } from '../sourceMaps';

const definition: SmartContractDefinition = {
  networks: {
    local: {
      address: 'Adiq4sMWabmjD7S45BCvW2uVgFDofmTbyo',
    },
  },
  abi: icoABI,
  sourceMaps,
};

export const createICOSmartContract = <TClient extends Client>(client: TClient): ICOSmartContract<TClient> =>
  client.smartContract<ICOSmartContract<TClient>>(definition);
