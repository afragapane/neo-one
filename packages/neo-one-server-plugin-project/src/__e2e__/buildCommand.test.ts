import {
  ABI,
  Client,
  Contract,
  DeveloperClient,
  Hash256,
  LocalKeyStore,
  LocalMemoryStore,
  LocalUserAccountProvider,
  NEOONEProvider,
  ReadClient,
  UserAccountID,
  wifToPrivateKey,
} from '@neo-one/client';
import { Network } from '@neo-one/server-plugin-network';
import { Wallet } from '@neo-one/server-plugin-wallet';
import { SetupTestResult } from '@neo-one/smart-contract-test';
import BigNumber from 'bignumber.js';
import * as path from 'path';
import { ICOReadSmartContract, ICOSmartContract } from '../__data__/ico/one/generated/ICO/types';

const getNetworks = async (): Promise<ReadonlyArray<Network>> => {
  const networksString = await one.execute('get network --json');

  return JSON.parse(networksString);
};

const getWallets = async (networkName: string): Promise<ReadonlyArray<Wallet>> => {
  const walletsString = await one.execute(`get wallet --json --network ${networkName}`);

  return JSON.parse(walletsString);
};

const LOCAL = 'local';

const getClients = async (
  network: Network,
  wallet: Wallet,
): Promise<{ readonly client: Client; readonly developerClient: DeveloperClient }> => {
  const keystore = new LocalKeyStore({ store: new LocalMemoryStore() });

  if (wallet.wif === undefined) {
    throw new Error(`Something went wrong, wif is null for ${wallet.name}`);
  }
  await keystore.addAccount({
    network: LOCAL,
    name: wallet.name,
    privateKey: wifToPrivateKey(wallet.wif),
  });

  const provider = new NEOONEProvider([{ network: LOCAL, rpcURL: network.nodes[0].rpcAddress }]);
  const localUserAccountProvider = new LocalUserAccountProvider({ keystore, provider });
  const client = new Client({ memory: localUserAccountProvider });

  const developerClient = new DeveloperClient(provider.read(LOCAL));

  return { client, developerClient };
};

const getContract = async (client: Client): Promise<Contract> => {
  const readClient = client.read(LOCAL);
  const blockHeight = await readClient.getBlockCount();

  // tslint:disable-next-line no-loop-statement
  for await (const block of readClient.iterBlocks({ indexStop: blockHeight })) {
    // tslint:disable-next-line no-loop-statement
    for (const transaction of block.transactions) {
      if (transaction.type === 'InvocationTransaction' && transaction.invocationData.contracts.length > 0) {
        return transaction.invocationData.contracts[0];
      }
    }
  }

  throw new Error('Could not find deployed contract');
};

const getGeneratedCode = (): {
  readonly abi: ABI;
  readonly contract: {
    readonly createSmartContract: (client: Client) => ICOSmartContract;
    readonly createReadSmartContract: (readClient: ReadClient) => ICOReadSmartContract;
  };
  readonly test: {
    readonly setupTest: () => Promise<SetupTestResult<ICOSmartContract>>;
  };
} => {
  // tslint:disable-next-line no-require-imports
  const abi = require('../__data__/ico/one/generated/ICO/abi').icoABI;
  // tslint:disable-next-line no-require-imports
  const contract = require('../__data__/ico/one/generated/ICO/contract');
  const createSmartContract = contract.createICOSmartContract;
  const createReadSmartContract = contract.createICOReadSmartContract;
  // tslint:disable-next-line no-require-imports
  const setupTest = require('../__data__/ico/one/generated/ICO/test').setupICOTest;

  return { abi, contract: { createSmartContract, createReadSmartContract }, test: { setupTest } };
};

const verifySmartContractAfterMint = async (
  smartContract: ICOSmartContract,
  accountID: UserAccountID,
): Promise<void> => {
  const [totalSupply, remaining, balance] = await Promise.all([
    smartContract.totalSupply(),
    smartContract.remaining(),
    smartContract.balanceOf(accountID.address),
  ]);
  expect(totalSupply.toString()).toEqual('100');
  expect(remaining.toString()).toEqual(new BigNumber(9_999_999_900).toString());
  expect(balance.toString()).toEqual('100');
};

const verifySmartContract = async (
  smartContract: ICOSmartContract,
  readSmartContract: ICOReadSmartContract,
  developerClient: DeveloperClient,
  accountID: UserAccountID,
): Promise<void> => {
  const [
    name,
    symbol,
    decimals,
    amountPerNEO,
    owner,
    startTimeSeconds,
    icoDurationSeconds,
    initialTotalSupply,
    initialRemaining,
    initialBalance,
  ] = await Promise.all([
    readSmartContract.name(),
    readSmartContract.symbol(),
    readSmartContract.decimals(),
    readSmartContract.amountPerNEO(),
    readSmartContract.owner(),
    readSmartContract.startTimeSeconds(),
    readSmartContract.icoDurationSeconds(),
    readSmartContract.totalSupply(),
    readSmartContract.remaining(),
    readSmartContract.balanceOf(accountID.address),
  ]);
  expect(name).toEqual('One');
  expect(symbol).toEqual('ONE');
  expect(decimals.toString()).toEqual('8');
  expect(amountPerNEO.toString()).toEqual('10');
  expect(owner).toEqual(accountID.address);
  expect(startTimeSeconds.toString()).toEqual('1534108415');
  expect(icoDurationSeconds.toString()).toEqual('157700000');
  expect(initialTotalSupply.toString()).toEqual('0');
  expect(initialRemaining.toString()).toEqual(new BigNumber(10_000_000_000).toString());
  expect(initialBalance.toString()).toEqual('0');

  const mintResult = await smartContract.mintTokens({
    transfers: [
      {
        to: smartContract.definition.networks[accountID.network].address,
        amount: new BigNumber(10),
        asset: Hash256.NEO,
      },
    ],
  });

  const [mintReceipt] = await Promise.all([mintResult.confirmed(), developerClient.runConsensusNow()]);
  if (mintReceipt.result.state === 'FAULT') {
    throw new Error(mintReceipt.result.message);
  }

  expect(mintReceipt.result.state).toEqual('HALT');
  expect(mintReceipt.result.value).toEqual(true);
  expect(mintReceipt.result.gasCost).toMatchSnapshot('mint cost');
  expect(mintReceipt.result.gasConsumed).toMatchSnapshot('mint consumed');
  expect(mintReceipt.events).toHaveLength(1);
  const event = mintReceipt.events[0];
  expect(event.name).toEqual('transfer');
  if (event.name !== 'transfer') {
    throw new Error('For TS');
  }
  expect(event.parameters.from).toBeUndefined();
  expect(event.parameters.to).toEqual(accountID.address);
  expect(event.parameters.amount.toString()).toEqual('100');

  await verifySmartContractAfterMint(smartContract, accountID);
};

describe('buildCommand', () => {
  test('build', async () => {
    await one.execute('build', { cwd: path.resolve(__dirname, '..', '__data__', 'ico') });

    const networks = await getNetworks();
    expect(networks.length).toEqual(1);

    const network = networks[0];
    expect(network.name).toEqual('ico-local');
    expect(network.type).toEqual('private');
    expect(network.nodes.length).toEqual(1);
    expect(network.nodes[0].live).toBeTruthy();
    expect(network.nodes[0].ready).toBeTruthy();

    const wallets = await getWallets(LOCAL);
    expect(wallets.length).toEqual(1);

    const wallet = wallets[0];
    const { client, developerClient } = await getClients(network, wallet);

    const contract = await getContract(client);
    expect(contract.codeVersion).toEqual('1.0');
    expect(contract.author).toEqual('dicarlo2');
    expect(contract.email).toEqual('alex.dicarlo@neotracker.io');
    expect(contract.description).toEqual('NEOâ€¢ONE ICO');
    expect(contract.parameters).toEqual(['String', 'Array']);
    expect(contract.returnType).toEqual('Buffer');
    expect(contract.payable).toBeTruthy();
    expect(contract.storage).toBeTruthy();
    expect(contract.dynamicInvoke).toBeFalsy();

    const {
      abi,
      contract: { createSmartContract, createReadSmartContract },
      test: { setupTest },
    } = getGeneratedCode();
    expect(abi).toBeDefined();

    const smartContract = createSmartContract(client);
    const readSmartContract = createReadSmartContract(client.read(LOCAL));
    const accountID = { ...wallet.accountID, network: LOCAL };
    const {
      developerClient: testDeveloperClient,
      smartContract: testSmartContract,
      masterAccountID,
      networkName,
      cleanup,
    } = await setupTest();
    const [deployResult] = await Promise.all([
      testSmartContract.deploy(masterAccountID.address, new BigNumber(1534108415), new BigNumber(157700000)),
      verifySmartContract(smartContract, readSmartContract, developerClient, accountID),
    ]);

    const [deployReceipt] = await Promise.all([deployResult.confirmed(), testDeveloperClient.runConsensusNow()]);
    if (deployReceipt.result.state === 'FAULT') {
      throw new Error(deployReceipt.result.message);
    }

    await verifySmartContract(
      testSmartContract,
      testSmartContract.read(networkName),
      testDeveloperClient,
      masterAccountID,
    );

    await one.execute('build', { cwd: path.resolve(__dirname, '..', '__data__', 'ico') });

    await verifySmartContractAfterMint(smartContract, accountID);

    await cleanup();
  });
});
