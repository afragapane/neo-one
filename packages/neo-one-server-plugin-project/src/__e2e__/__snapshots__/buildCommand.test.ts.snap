// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCommand build: mint consumed 1`] = `"0"`;

exports[`buildCommand build: mint consumed 2`] = `"0"`;

exports[`buildCommand build: mint cost 1`] = `"6.659"`;

exports[`buildCommand build: mint cost 2`] = `"6.659"`;
